### PR TITLE
[Feature][Part 1]Library rework - Media pages

### PR DESCRIPTION
**Pull request by Mikunj**
This contains the changes for the media pages. It also has the foundations laid out for the second part which is the library page itself.

**Changes:**

- Added entry syncing between library, media pages and quick update.
- Updated media page to match the new design.
- QoL fixes.

### CODE REVIEW

File: src/screens/Profiles/MediaPages/index.js [hunk 1]
```diff
@@ -70,21 +73,31 @@ class MediaPages extends PureComponent {
...
       case 'on_hold':
       case 'dropped':
-        libraryEntry ? await this.updateLibraryEntry(option) : await this.createLibraryEntry(option);
+        const data = { state: option };
+        libraryEntry ? await this.updateLibraryEntry(data) : await this.createLibraryEntry(data);
         break;
       case 'remove':
         this.setState({ loadingLibrary: true });
         await Kitsu.destroy('libraryEntries', libraryEntry.id);
+        KitsuLibrary.onLibraryEntryDelete(libraryEntry.id, mediaType, libraryEntry.status, 'mediapage');
```
Comment (vevix) [inline src/screens/Profiles/MediaPages/index.js]: We should maybe create an enum-type object for source?

File: src/screens/Profiles/MediaPages/index.js [hunk 2]
```diff
@@ -18,13 +18,16 @@ import { coverImageHeight, scene } from 'kitsu/screens/Profiles/constants';
...
+import { KitsuLibrary, KitsuLibraryEvents } from 'kitsu/utils/kitsuLibrary';
 
 const HEADER_HEIGHT = navigationBarHeight + statusBarHeight + (isX ? paddingX : 0);
 const TAB_ITEMS = [
   { key: 'summary', label: 'Summary', screen: 'Summary' },
   { key: 'episodes', label: 'Episodes', screen: 'Episodes', if: (state) => state.media.type === 'anime'},
   { key: 'chapters', label: 'Chapters', screen: 'Episodes', if: (state) => state.media.type === 'manga'},
-  { key: 'characters', label: 'Characters', screen: 'Characters' },
+  // NOTE: Disabled until we improve char db
+  // { key: 'characters', label: 'Characters', screen: 'Characters' },
```
Comment (vevix) [inline src/screens/Profiles/MediaPages/index.js]: We should disable the network request as well.

File: src/screens/Profiles/MediaPages/index.js [hunk 3]
```diff
@@ -120,40 +133,146 @@ class MediaPages extends PureComponent {
...
+
+    // Check progress is within the bounds
+    let progress = number;
+    if (progress < 0) progress = 0;
+    if (mediaCount && progress > mediaCount) {
+      progress = mediaCount;
+    }
+    changes = { ...changes, progress };
+
+    // Mark entry as completed if the media has finished airing and the progress is the same as the count.
```
Comment (vevix) [inline src/screens/Profiles/MediaPages/index.js]: We might want to not check the media status here, Could hit an issue where DB might be slower to update than the final episode to air.

File: src/screens/Profiles/MediaPages/index.js [hunk 4]
```diff
@@ -70,21 +73,31 @@ class MediaPages extends PureComponent {
...
     const { libraryEntry } = this.state;
+    const { mediaType } = this.props.navigation.state.params;
     switch (option) {
       case 'current':
       case 'planned':
       case 'completed':
       case 'on_hold':
       case 'dropped':
-        libraryEntry ? await this.updateLibraryEntry(option) : await this.createLibraryEntry(option);
+        const data = { state: option };
```
Comment (vevix) [inline src/screens/Profiles/MediaPages/index.js]: `status`

File: src/screens/Profiles/MediaPages/pages/Summary/progress.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  static defaultProps = {
+    media: null,
+    libraryEntry: null,
+    onPress: null,
+    onEditPress: null,
+  }
+
+  navigateToEdit = (props) => {
+    this.props.navigation.navigate('PostDetails', props);
+  }
```
Comment (vevix) [inline src/screens/Profiles/MediaPages/pages/Summary/progress.js]: Not used? and wrong route?

File: src/screens/Profiles/ProfilePages/index.js [hunk 6]
```diff
@@ -305,8 +306,19 @@ class ProfilePage extends PureComponent {
...
+          <CustomHeader
+            leftButtonAction={this.goBack}
+            leftButtonTitle="Back"
+          />
+          <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
+            <StyledText color="light" textStyle={{ textAlignVertical: 'center', textAlign: 'center' }}>
+              An Error Occured.
+            </StyledText>
+          </View>
+        </SceneContainer>
```
Comment (vevix) [inline src/screens/Profiles/ProfilePages/index.js]: Maybe we should pull this out to a component.

File: src/utils/kitsuLibrary.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  constructor() {
+    this.handlers = [];
+  }
+
+  /**
+   * Subscribe to a library event.
+   *
+   * @param {any} event The event to subscribe to. Can find them in `KitsuLibraryEvents` class.
+   * @param {any} handler The handler function which takes in 1 argument.
+   * @returns The unsubsribe function.
```
Comment (vevix) [inline src/utils/kitsuLibrary.js]: `unsubscribe`

File: src/screens/Profiles/MediaPages/index.js (hunk 3 above)
Comment (Mikunj) [reply src/screens/Profiles/MediaPages/index.js]: hmm, what should i change the logic to? just check if progress is the same as the count?

File: src/screens/Profiles/UserLibrary/components/UserLibraryEditScreen/component.js [hunk 8]
```diff
@@ -145,169 +154,203 @@ export class UserLibraryEditScreenComponent extends React.Component {
...
-                {status}
+          {/* Loading */}
+          {loading &&
+            <View style={styles.loadingContainer}>
+              <ActivityIndicator size="large" color="white" />
+            </View>
+          }
+
+          {/* Error */}
+          {error &&
```
Comment (edencakir) [inline src/screens/Profiles/UserLibrary/components/UserLibraryEditScreen/component.js]: if `error` is `''` then we'll have an of strings should be placed inside <Text> etc. Let's change it to `isNull(error)` from `lodash`; or something similar

File: src/screens/Profiles/UserLibrary/components/UserLibraryEditScreen/component.js [hunk 9]
```diff
@@ -145,169 +154,203 @@ export class UserLibraryEditScreenComponent extends React.Component {
...
+                      {moment(this.state.startedAt).format('MMM. DD, YYYY')}
+                    </Text>
+                  }
+                </View>
+              </View>
+              <View style={[styles.editRow, styles.dateFinished]}>
+                <View>
+                  <Text style={[
+                    styles.editRowLabel,
+                    this.state.finishedAt && styles.withValueLabel,
```
Comment (edencakir) [inline src/screens/Profiles/UserLibrary/components/UserLibraryEditScreen/component.js]: we can deconstruct `finishedAt` and `startedAt` before rendering to improve readability:

`const { finishedAt } = this.state` like you did above.

File: src/screens/Profiles/UserLibrary/components/UserLibraryEditScreen/styles.js [hunk 10]
```diff
@@ -7,6 +7,25 @@ export const styles = StyleSheet.create({
...
+  loadingContainer: {
+    position: 'absolute',
+    top: 0,
+    left: 0,
+    bottom: 0,
+    right: 0,
+    backgroundColor: 'rgba(0,0,0,0.7)',
+    alignItems: 'center',
+    justifyContent: 'center',
+    zIndex: 666,
```
Comment (edencakir) [inline src/screens/Profiles/UserLibrary/components/UserLibraryEditScreen/styles.js]: Should we export these indexes to layer => into a constants file.

Just a thought.

File: src/screens/Profiles/components/ErrorPage/component.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+import { StyledText } from 'kitsu/components/StyledText';
+import { CustomHeader } from 'kitsu/screens/Profiles/components/CustomHeader';
+
+export const ErrorPage = ({ errorText, onBackPress }) => (
+  <SceneContainer>
+    <CustomHeader
+      leftButtonAction={onBackPress}
+      leftButtonTitle="Back"
+    />
+    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
```
Comment (edencakir) [inline src/screens/Profiles/components/ErrorPage/component.js]: Let's move them to `./styles`

File: src/screens/Profiles/components/SceneHeader/component.js [hunk 12]
```diff
@@ -110,15 +124,21 @@ export class SceneHeader extends PureComponent {
...
     const {
       variant,
-      type,
       title,
       posterImage,
       moreButtonOptions,
       onMoreButtonOptionsSelected,
       showMoreButton,
-      subType,
+      subTitle,
```
Comment (edencakir) [inline src/screens/Profiles/components/SceneHeader/component.js]: Same as above, subtitle?

File: src/screens/Profiles/MediaPages/index.js (hunk 3 above)
Comment (vevix) [reply src/screens/Profiles/MediaPages/index.js]: Yeah, that's all web does.

File: src/screens/Profiles/MediaPages/pages/Summary/progress.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    onEditPress: null,
+  }
+
+  render() {
+    const { libraryEntry, media, onPress, onEditPress } = this.props;
+
+    const completed = libraryEntry && libraryEntry.status === 'completed';
+    const mediaCount = media && (media.episodeCount || media.chapterCount);
+
+    const progress = (libraryEntry && mediaCount && (libraryEntry.progress || 0) / mediaCount) || 0;
```
[the remark this answers is not in the document]
Comment (Mikunj) [reply src/screens/Profiles/MediaPages/pages/Summary/progress.js]: We apply a `Math.floor` below it

**Pull request merged**